Error while generating a new configuration with labels

Hi everyone,
before the error's description there is my setup:

- BSB branch: fix/labelled_ps
- cerebellar_models branch: feature/plasticity
- yaml file used: `circuit.yaml`, alredy available in the branch. generated by the API but modified afterwards
- command line used: `mpirun -n 8 bsb compile circuit.yaml --clear`


Also, I already updated feature/plasticity to be on par with the master branch, except for the recent commits to update `io_molecular.py` and all the custom plasticity section.
About the error, the label seem to work correclty now, which means that the IO labelled 'plus' connect with the PC labelled 'plus' (considering the synapse io_to_purkinje_plus), etc.

But now, the issue goes to the io_to_mli connectivity, both plus and minus, with the following error (this is from the connectivty io_to_mli_plus):
 ```
File "/home/emiliano/Documenti/other_env/cerebellar-models/cerebellar_models/connectome/io_molecular.py", line 173, in connect
    |     self.connect_cells(
    |   File "/home/emiliano/Documenti/other_env/bsb/packages/bsb-core/bsb/connectivity/strategy.py", line 244, in connect_cells
    |     self.scaffold.connect_cells(pre_set, post_set, src_locs, dest_locs, name)
    |   File "/home/emiliano/Documenti/other_env/bsb/packages/bsb-core/bsb/core.py", line 606, in connect_cells
    |     cs.connect(pre_set, post_set, src_locs, dest_locs)
    |   File "/home/emiliano/Documenti/other_env/bsb/packages/bsb-hdf5/bsb_hdf5/resource.py", line 49, in handle_indirection
    |     return f(*bound.args, **bound.kwargs)
    |   File "/home/emiliano/Documenti/other_env/bsb/packages/bsb-hdf5/bsb_hdf5/connectivity_set.py", line 169, in connect
    |     src_locs[:, 0] = mask_labels[src_locs[:, 0]]
    | IndexError: index 12 is out of bounds for axis 0 with size 7
```

Moreover, this 12 specified in the error is one of the ids in the IO labelled as 'plus'.

And, to conclude, before compilng the yaml file, please comment all the simulations section. At the moment it is not the problem and the custom synapse needs to be installed, which is not necessary for this issue.

i think it compiled correctly? exit code 0. not sure what `osc_ucx_component.c:375  Error: OSC UCX component priority set inside component query failed` are about. I'll try loading from it and get back to you.

<details><summary>Complete output</summary>
<p>

```txt
~/Documents/work/poli/robocontroller/compile-hdf5/cerebellar-models feature/plasticity !1 ❮ mpirun -np 4 bsb compile -v3 circuit.yaml                           (compile-hdf5)  󰌠 3.10.0
/home/paolo/programs/micromamba/envs/compile-hdf5/lib/python3.10/site-packages/requests/__init__.py:113: RequestsDependencyWarning: urllib3 (2.6.3) or chardet (7.0.0)/charset_normalizer (3.4.4) doesn't match a supported version!
  warnings.warn(
/home/paolo/programs/micromamba/envs/compile-hdf5/lib/python3.10/site-packages/requests/__init__.py:113: RequestsDependencyWarning: urllib3 (2.6.3) or chardet (7.0.0)/charset_normalizer (3.4.4) doesn't match a supported version!
  warnings.warn(
/home/paolo/programs/micromamba/envs/compile-hdf5/lib/python3.10/site-packages/requests/__init__.py:113: RequestsDependencyWarning: urllib3 (2.6.3) or chardet (7.0.0)/charset_normalizer (3.4.4) doesn't match a supported version!
  warnings.warn(
/home/paolo/programs/micromamba/envs/compile-hdf5/lib/python3.10/site-packages/requests/__init__.py:113: RequestsDependencyWarning: urllib3 (2.6.3) or chardet (7.0.0)/charset_normalizer (3.4.4) doesn't match a supported version!
  warnings.warn(
/home/paolo/Documents/work/poli/robocontroller/compile-hdf5/bsb/packages/bsb-core/bsb/_package_spec.py:54: PackageRequirementWarning: Missing package 'cerebellar-models'. You may experience errors or differences in results.
  warn(warning, PackageRequirementWarning)
/home/paolo/Documents/work/poli/robocontroller/compile-hdf5/bsb/packages/bsb-core/bsb/_package_spec.py:54: PackageRequirementWarning: Missing package 'bsb'. You may experience errors o
r differences in results.
  warn(warning, PackageRequirementWarning)
/home/paolo/Documents/work/poli/robocontroller/compile-hdf5/bsb/packages/bsb-core/bsb/_package_spec.py:54: PackageRequirementWarning: Missing package 'cerebellar-models'. You may experience errors or differences in results.
  warn(warning, PackageRequirementWarning)
/home/paolo/Documents/work/poli/robocontroller/compile-hdf5/bsb/packages/bsb-core/bsb/_package_spec.py:54: PackageRequirementWarning: Missing package 'bsb'. You may experience errors or differences in results.
  warn(warning, PackageRequirementWarning)
/home/paolo/Documents/work/poli/robocontroller/compile-hdf5/bsb/packages/bsb-core/bsb/_package_spec.py:54: PackageRequirementWarning: Missing package 'cerebellar-models'. You may experience errors or differences in results.
  warn(warning, PackageRequirementWarning)
/home/paolo/Documents/work/poli/robocontroller/compile-hdf5/bsb/packages/bsb-core/bsb/_package_spec.py:54: PackageRequirementWarning: Missing package 'bsb'. You may experience errors or differences in results.
  warn(warning, PackageRequirementWarning)
/home/paolo/Documents/work/poli/robocontroller/compile-hdf5/bsb/packages/bsb-core/bsb/_package_spec.py:54: PackageRequirementWarning: Missing package 'cerebellar-models'. You may experience errors or differences in results.
  warn(warning, PackageRequirementWarning)
/home/paolo/Documents/work/poli/robocontroller/compile-hdf5/bsb/packages/bsb-core/bsb/_package_spec.py:54: PackageRequirementWarning: Missing package 'bsb'. You may experience errors o
r differences in results.
  warn(warning, PackageRequirementWarning)
[pc-pp-fw:62427] osc_ucx_component.c:375  Error: OSC UCX component priority set inside component query failed

[pc-pp-fw:62424] osc_ucx_component.c:375  Error: OSC UCX component priority set inside component query failed

[pc-pp-fw:62425] osc_ucx_component.c:375  Error: OSC UCX component priority set inside component query failed

[pc-pp-fw:62426] osc_ucx_component.c:375  Error: OSC UCX component priority set inside component query failed

[pc-pp-fw:62425] osc_ucx_component.c:375  Error: OSC UCX component priority set inside component query failed

[pc-pp-fw:62426] osc_ucx_component.c:375  Error: OSC UCX component priority set inside component query failed

[pc-pp-fw:62427] osc_ucx_component.c:375  Error: OSC UCX component priority set inside component query failed

[pc-pp-fw:62424] osc_ucx_component.c:375  Error: OSC UCX component priority set inside component query failed

Compiling the following strategies: granular_layer_innervation, granular_layer_placement, purkinje_layer_placement, basket_layer_placement, stellate_layer_placement, dcn_p_layer_placement, dcn_i_layer_placement, io_layer_placement, mossy_fibers_to_glomerulus, glomerulus_to_granule, glomerulus_to_golgi, golgi_to_glomerulus, gap_goc, golgi_to_golgi, ascending_axon_to_golgi, parallel_fiber_to_golgi, ascending_axon_to_purkinje_plus, ascending_axon_to_purkinje_minus, parallel_fiber_to_purkinje_plus, parallel_fiber_to_purkinje_minus, parallel_fiber_to_basket, parallel_fiber_to_stellate, basket_to_purkinje_plus, basket_to_purkinje_minus, stellate_to_purkinje_plus, stellate_to_purkinje_minus, basket_to_basket, stellate_to_stellate, purkinje_to_dcn_p_plus, purkinje_to_dcn_p_minus, purkinje_to_dcn_i_plus, purkinje_to_dcn_i_minus, mossy_fibers_to_dcn_p_plus, mossy_fibers_to_dcn_p_minus, io_to_purkinje_plus, io_to_purkinje_minus, io_to_mli_plus, io_to_mli_minus, io_to_dcn_p_plus, io_to_dcn_p_minus, io_to_dcn_i_plus, io_to_dcn_i_minus, dcn_i_to_io_plus, dcn_i_to_io_minus
[pc-pp-fw:62424] osc_ucx_component.c:375  Error: OSC UCX component priority set inside component query failed

[pc-pp-fw:62427] osc_ucx_component.c:375  Error: OSC UCX component priority set inside component query failed

[pc-pp-fw:62425] osc_ucx_component.c:375  Error: OSC UCX component priority set inside component query failed

[pc-pp-fw:62426] osc_ucx_component.c:375  Error: OSC UCX component priority set inside component query failed

Starting placement strategies: granular_layer_innervation, granular_layer_placement, purkinje_layer_placement, basket_layer_placement, stellate_layer_placement, dcn_p_layer_placement, dcn_i_layer_placement, io_layer_placement
[pc-pp-fw:62424] osc_ucx_component.c:375  Error: OSC UCX component priority set inside component query failed

[pc-pp-fw:62426] osc_ucx_component.c:375  Error: OSC UCX component priority set inside component query failed

[pc-pp-fw:62425] osc_ucx_component.c:375  Error: OSC UCX component priority set inside component query failed

[pc-pp-fw:62427] osc_ucx_component.c:375  Error: OSC UCX component priority set inside component query failed

[pc-pp-fw:62424] osc_ucx_component.c:375  Error: OSC UCX component priority set inside component query failed

[pc-pp-fw:62427] osc_ucx_component.c:375  Error: OSC UCX component priority set inside component query failed

[pc-pp-fw:62425] osc_ucx_component.c:375  Error: OSC UCX component priority set inside component query failed

[pc-pp-fw:62426] osc_ucx_component.c:375  Error: OSC UCX component priority set inside component query failed

Starting connectivity strategies: mossy_fibers_to_glomerulus, glomerulus_to_granule, glomerulus_to_golgi, golgi_to_glomerulus, gap_goc, golgi_to_golgi, ascending_axon_to_golgi, parallel_fiber_to_golgi, ascending_axon_to_purkinje_plus, ascending_axon_to_purkinje_minus, parallel_fiber_to_purkinje_plus, parallel_fiber_to_purkinje_minus, parallel_fiber_to_basket, parallel_fiber_to_stellate, basket_to_purkinje_plus, basket_to_purkinje_minus, stellate_to_purkinje_plus, stellate_to_purkinje_minus, basket_to_basket, stellate_to_stellate, purkinje_to_dcn_p_plus, purkinje_to_dcn_p_minus, purkinje_to_dcn_i_plus, purkinje_to_dcn_i_minus, mossy_fibers_to_dcn_p_plus, mossy_fibers_to_dcn_p_minus, io_to_purkinje_plus, io_to_purkinje_minus, io_to_mli_plus, io_to_mli_minus, io_to_dcn_p_plus, io_to_dcn_p_minus, io_to_dcn_i_plus, io_to_dcn_i_minus, dcn_i_to_io_plus, dcn_i_to_io_minus
[pc-pp-fw:62424] osc_ucx_component.c:375  Error: OSC UCX component priority set inside component query failed

[pc-pp-fw:62427] osc_ucx_component.c:375  Error: OSC UCX component priority set inside component query failed

[pc-pp-fw:62425] osc_ucx_component.c:375  Error: OSC UCX component priority set inside component query failed

[pc-pp-fw:62426] osc_ucx_component.c:375  Error: OSC UCX component priority set inside component query failed

[pc-pp-fw:62424] osc_ucx_component.c:375  Error: OSC UCX component priority set inside component query failed

```

</p>
</details>

Happy to hear that it seems to have solved the issue. 

> `osc_ucx_component.c:375  Error: OSC UCX component priority set inside component query failed`
> installing this version messes with my MPI install 

It would be difficult for me to help you on these unless you give me more information on how you installed both the bsb and the cerebellar-models packages. It may well be that your MPI installation and UCX issues are linked. BSB requires certain c++ library for mpi and these might conflict with UCX?  
>  do you need a PR or anything else from my side?

We will make sure that this fix is integrated to the next release of cerebellar-models . To be more precise, [this branch](https://github.com/dbbs-lab/cerebellar-models/tree/feature/new_microzones) is the most likely to be merged on master the soonest, so we will probably add the fix to it.

--- a/.github/devops/install-nest.sh
+++ b/.github/devops/install-nest.sh
@@ -3,7 +3,7 @@ cd $GITHUB_WORKSPACE/_nest_repo
 git checkout tags/v$1
 mkdir build
 cd build
-pip install cython~=3.0.0
+pip install cython~=3.0.0 cmake
 mkdir -p $2
 cmake .. \
   -DCMAKE_INSTALL_PREFIX=$2 \

--- a/.github/workflows/integration.yml
+++ b/.github/workflows/integration.yml
@@ -32,15 +32,6 @@ jobs:
         # Install `libopenmpi` for MPI and libgsl for NEST
         sudo apt install openmpi-bin libopenmpi-dev libgsl-dev
 
-    - name: Cache pip
-      uses: actions/cache@v4
-      with:
-        path: ~/.cache/pip
-        key: ${{ runner.os }}-pip-${{ hashFiles('pyproject.toml') }}
-        restore-keys: |
-          ${{ runner.os }}-pip-
-          ${{ runner.os }}-
-
     - name: Install dependencies & self
       run: |
         pip install --upgrade pip

--- a/.github/workflows/integration.yml
+++ b/.github/workflows/integration.yml
@@ -58,6 +58,9 @@ jobs:
       run: .github/devops/install-nest.sh ${NEST_VERSION} ${{ github.workspace }}/simulators/nest
 
     - name: Run tests & coverage
+      timeout-minutes: 60
+      env:
+        BSB_QUIET: true
       run: |
         source ${{ github.workspace }}/simulators/nest/bin/nest_vars.sh
         coverage run -p -m unittest discover -v -s ./tests

--- a/.github/workflows/integration.yml
+++ b/.github/workflows/integration.yml
@@ -32,17 +32,6 @@ jobs:
         # Install `libopenmpi` for MPI and libgsl for NEST
         sudo apt install openmpi-bin libopenmpi-dev libgsl-dev
 
-    - name: Cache NEST
-      id: cache-nest
-      uses: actions/cache@v4
-      with:
-        path: ${{ github.workspace }}/simulators/nest
-        key: ${{ runner.os }}-${{ matrix.python-version }}-nest
-
-    - name: Install NEST
-      if: steps.cache-nest.outputs.cache-hit != 'true'
-      run: .github/devops/install-nest.sh ${NEST_VERSION} ${{ github.workspace }}/simulators/nest
-
     - name: Cache pip
       uses: actions/cache@v4
       with:
@@ -56,6 +45,17 @@ jobs:
       run: |
         pip install --upgrade pip
         pip install -e .[test]
+
+    - name: Cache NEST
+      id: cache-nest
+      uses: actions/cache@v4
+      with:
+        path: ${{ github.workspace }}/simulators/nest
+        key: ${{ runner.os }}-${{ matrix.python-version }}-nest
+
+    - name: Install NEST
+      if: steps.cache-nest.outputs.cache-hit != 'true'
+      run: .github/devops/install-nest.sh ${NEST_VERSION} ${{ github.workspace }}/simulators/nest
 
     - name: Run tests & coverage
       run: |

--- a/cerebellar_models/analysis/plots.py
+++ b/cerebellar_models/analysis/plots.py
@@ -195,12 +195,23 @@ class ScaffoldPlot(Plot):
         extra = "_".join(labels)
         return ct_name + (extra if len(extra) == 0 else "_" + extra)
 
+    @staticmethod
+    def get_unique_labels(ps):
+        """
+        Get unique labels sets for a provided placement set
+        Provide the default placement set label sets if the placement set is empty.
+        """
+        u_labels = ps.get_unique_labels()
+        if not len(u_labels):
+            u_labels = [set()]
+        return u_labels
+
     @property
     def labelled_dict_colors(self):
         result = self.dict_colors.copy()
         for ps in self.scaffold.get_placement_sets():
             ct_name = ps.cell_type.name
-            u_labels = ps.get_unique_labels()
+            u_labels = self.get_unique_labels(ps)
             if len(u_labels) > 1:
                 color = np.array(result[ct_name])
                 del result[ct_name]
@@ -236,9 +247,10 @@ class Legend(Plot):
             patches.Patch(color=color[:3], label=label) for label, color in self.dict_colors.items()
         ]
         dict_plot = self.dict_legend.copy()
+        dict_plot["ncol"] = self.cols_legend
         dict_plot.update(kwargs)
         keys = [self.dict_abbreviations.get(k, k) for k in self.dict_colors.keys()]
-        self.get_ax().legend(patchs, keys, ncol=self.cols_legend, **dict_plot)
+        self.get_ax().legend(patchs, keys, **dict_plot)
 
     def remove_ct(self, to_keep: List[str], to_ignore: List[str] = None):
         """

--- a/cerebellar_models/analysis/report.py
+++ b/cerebellar_models/analysis/report.py
@@ -169,7 +169,7 @@ class BSBReport(Report):
         ct_names = []
         for ps in self.scaffold.get_placement_sets():
             ct_name = ps.cell_type.name
-            for labels in ps.get_unique_labels():
+            for labels in ScaffoldPlot.get_unique_labels(ps):
                 ct_names.append(ScaffoldPlot.get_labelled_ct_name(ct_name, labels))
         return ct_names
 

--- a/cerebellar_models/analysis/spiking_results.py
+++ b/cerebellar_models/analysis/spiking_results.py
@@ -753,10 +753,10 @@ class SimResultsTable(TablePlot, SpikePlot):
             spikes = loc_spikes[i]
 
             unique_counts = np.unique(spikes.array_annotations["senders"], return_counts=True)[1]
-            unique_counts = np.concatenate([unique_counts, np.zeros(self.nb_neurons[i]-len(unique_counts))])
-            all_fr = unique_counts /  (
-                (self.time_to - self.time_from) / 1000.0
+            unique_counts = np.concatenate(
+                [unique_counts, np.zeros(self.nb_neurons[i] - len(unique_counts))]
             )
+            all_fr = unique_counts / ((self.time_to - self.time_from) / 1000.0)
             isi = extract_isis(spikes, self.dt)
 
             self._values.append([all_fr, isi])

--- a/cerebellar_models/analysis/spiking_results.py
+++ b/cerebellar_models/analysis/spiking_results.py
@@ -752,7 +752,9 @@ class SimResultsTable(TablePlot, SpikePlot):
         for i in range(num_filter):
             spikes = loc_spikes[i]
 
-            all_fr = np.unique(spikes.array_annotations["senders"], return_counts=True)[1] / (
+            unique_counts = np.unique(spikes.array_annotations["senders"], return_counts=True)[1]
+            unique_counts = np.concatenate([unique_counts, np.zeros(self.nb_neurons[i]-len(unique_counts))])
+            all_fr = unique_counts /  (
                 (self.time_to - self.time_from) / 1000.0
             )
             isi = extract_isis(spikes, self.dt)

--- a/cerebellar_models/analysis/structure_analysis.py
+++ b/cerebellar_models/analysis/structure_analysis.py
@@ -99,7 +99,8 @@ class PlacementTable(TablePlot, ScaffoldPlot):
             ct = ps.cell_type
             volume = np.sum([p.volume() for place in ct.get_placement() for p in place.partitions])
             ct_name = self.extract_ct_name(ct)
-            for labels in ps.get_unique_labels():
+            u_labels = self.get_unique_labels(ps)
+            for labels in u_labels:
                 count = ps.get_labelled(labels).size
                 self.rows.append(self.get_labelled_ct_name(ct_name, labels))
                 self._values.append([count, volume])
@@ -351,7 +352,7 @@ class CellPlacement3D(ScaffoldPlot):
             if ct.name not in self.ignored_ct:
                 positions = ps.load_positions()
                 if len(positions) > 0:
-                    u_labels = ps.get_unique_labels()
+                    u_labels = self.get_unique_labels(ps)
                     if len(u_labels) > 1:
                         color = [
                             self.labelled_dict_colors.get(
@@ -505,10 +506,10 @@ class StructureReport(BSBReport):
         super().__init__(scaffold, cell_type_info)
         to_ignore = ["glomerulus", "ubc_glomerulus"]
         num_labelled_ct = sum(
-            len(ps.get_unique_labels()) for ps in self.scaffold.get_placement_sets()
+            len(ScaffoldPlot.get_unique_labels(ps)) for ps in self.scaffold.get_placement_sets()
         )
         legend = Legend(
-            (10, 0.6 * (num_labelled_ct - len(to_ignore)) / 3.0),
+            (10, 0.6 * max(num_labelled_ct - len(to_ignore), 1) / 3.0),
             3,
             dict_legend=dict(columnspacing=2.0, handletextpad=0.1, fontsize=20, loc="lower center"),
             dict_abbreviations=self.labelled_abbreviations,

--- a/cerebellar_models/cli.py
+++ b/cerebellar_models/cli.py
@@ -223,9 +223,10 @@ def _add_microzones(configuration, micro_params: MicrozonesParams):
     # add labelling strat
     configuration["after_placement"] = {
         "label_microzones": {
-            "strategy": "cerebellar_models.placement.microzones.LabelMicrozones",
+            "strategy": "cerebellar_models.placement.microzones.LabelCells",
             "cell_types": cells_found,
             "labels": micro_params.labels,
+            "same_size": True,
         }
     }
 

--- a/cerebellar_models/connectome/io_molecular.py
+++ b/cerebellar_models/connectome/io_molecular.py
@@ -166,6 +166,7 @@ class ConnectomeIO_MLI(NotParallel, ConnectionStrategy):
                     ptr = ptr + ln
 
             # because we are using global indices we need to extract the global ps
+            pre_ps = self.scaffold.get_placement_set(pre_ps.cell_type)
             for i, post_ps in enumerate(post.placement):
                 post_ps = self.scaffold.get_placement_set(post_ps.cell_type)
                 current = (ps_locs == i)[:ptr]

--- a/cerebellar_models/placement/microzones.py
+++ b/cerebellar_models/placement/microzones.py
@@ -27,7 +27,7 @@ class LabelCells(AfterPlacementHook):
             )
             return np.split(index_pos, split_indexes)[:-1]
         choice = np.random.choice(len(self.labels), len(ps))
-        return [np.where(choice==i)[0] for i in range(len(self.labels))]
+        return [np.where(choice == i)[0] for i in range(len(self.labels))]
 
     def postprocess(self):
         for cell_type in self.cell_types:
@@ -61,4 +61,3 @@ class LabelMicrozones(LabelCells):
             dtype=int,
         )
         return np.split(index_pos, split_indexes)[:-1]
-

--- a/cerebellar_models/placement/microzones.py
+++ b/cerebellar_models/placement/microzones.py
@@ -5,8 +5,7 @@ from bsb import AfterPlacementHook, config, refs, types
 @config.node
 class LabelCells(AfterPlacementHook):
     """
-    Subdivide cell populations into labelled subpopulations of
-    same cell counts randomly.
+    Subdivide cell populations into labelled subpopulations randomly.
     The number of labels defines the number of subpopulations
     """
 

--- a/cerebellar_models/placement/microzones.py
+++ b/cerebellar_models/placement/microzones.py
@@ -3,37 +3,62 @@ from bsb import AfterPlacementHook, config, refs, types
 
 
 @config.node
-class LabelMicrozones(AfterPlacementHook):
+class LabelCells(AfterPlacementHook):
     """
     Subdivide cell populations into labelled subpopulations of
-    same cell counts based on their position along a provided axis.
+    same cell counts randomly.
     The number of labels defines the number of subpopulations
     """
 
     cell_types: str = config.reflist(refs.cell_type_ref, required=True)
     """Reference to the cell type."""
 
-    axis: int = config.attr(type=types.int(min=0, max=2), default=0)
-    """Axis along which to subdivide the population."""
-
     labels: list[str] = config.list(type=str, default=["type1", "type2"])
     """List of labels to assign to each subpopulation."""
+
+    same_size: bool = config.attr(type=bool, default=False)
+
+    def split_indexes(self, ps):
+        if self.same_size:
+            index_pos = np.random.permutation(len(ps))
+            split_indexes = np.asarray(
+                np.round(np.linspace(0, len(index_pos), len(self.labels) + 1))[1:],
+                dtype=int,
+            )
+            return np.split(index_pos, split_indexes)[:-1]
+        choice = np.random.choice(len(self.labels), len(ps))
+        return [np.where(choice==i)[0] for i in range(len(self.labels))]
 
     def postprocess(self):
         for cell_type in self.cell_types:
             # Load the cell type positions
             ps = self.scaffold.get_placement_set(cell_type)
-            cell_positions = ps.load_positions()
+            split_indexes = self.split_indexes(ps)
+            for indexes, label in zip(split_indexes, self.labels):
+                ps.label(labels=[label], cells=indexes)
 
-            # create a filter that split the cells according to
-            # the mean of their positions along the chosen axis
-            index_pos = np.argsort(cell_positions[:, self.axis])
-            split_indexes = np.asarray(
-                np.round(np.linspace(0, len(index_pos), len(self.labels) + 1))[1:], dtype=int
-            )
-            last_i = 0
-            for i, label in zip(split_indexes, self.labels):
-                subpopulation_1 = np.asarray(index_pos[last_i:i], dtype=int)
-                # set the cell label according to the filter
-                ps.label(labels=[label], cells=subpopulation_1)
-                last_i = i
+
+@config.node
+class LabelMicrozones(LabelCells):
+    """
+    Subdivide cell populations into labelled subpopulations of
+    same cell counts based on their position along a provided axis.
+    The number of labels defines the number of subpopulations
+    """
+
+    axis: int = config.attr(type=types.int(min=0, max=2), default=0)
+    """Axis along which to subdivide the population."""
+
+    same_size = config.unset()
+
+    def split_indexes(self, ps):
+        cell_positions = ps.load_positions()
+        # create a filter that split the cells according to
+        # the mean of their positions along the chosen axis
+        index_pos = np.argsort(cell_positions[:, self.axis])
+        split_indexes = np.asarray(
+            np.round(np.linspace(0, len(index_pos), len(self.labels) + 1))[1:],
+            dtype=int,
+        )
+        return np.split(index_pos, split_indexes)[:-1]
+

--- a/cerebellar_models/placement/microzones.py
+++ b/cerebellar_models/placement/microzones.py
@@ -17,6 +17,7 @@ class LabelCells(AfterPlacementHook):
     """List of labels to assign to each subpopulation."""
 
     same_size: bool = config.attr(type=bool, default=False)
+    """Flag to split in ensembles of same size"""
 
     def split_indexes(self, ps):
         if self.same_size:

--- a/docs/strategies/placement-strategies.rst
+++ b/docs/strategies/placement-strategies.rst
@@ -1,6 +1,20 @@
 Placement strategies
 --------------------
 
+:class:`LabelCells<.placement.microzones.LabelCells>`
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+BSB :class:`AfterPlacementHook<bsb.postprocessing.AfterPlacementHook>` to subdivide cell
+populations into labelled subpopulations of same cell counts randomly.
+The number of labels defines the number of subpopulations.
+
+* ``cell_types``: List of CellType of the population to subdivide, used as references;
+
+* ``same_size``: *(bool, optional)* If `True`, the population is divided into equally sized subsets.
+  Default: `False`.
+
+* ``labels``: List of labels to assign to each population respectively, defines also the number of subpopulations;
+
+
 :class:`LabelMicrozones<.placement.microzones.LabelMicrozones>`
 ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
 

--- a/docs/strategies/placement-strategies.rst
+++ b/docs/strategies/placement-strategies.rst
@@ -4,16 +4,15 @@ Placement strategies
 :class:`LabelCells<.placement.microzones.LabelCells>`
 ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
 BSB :class:`AfterPlacementHook<bsb.postprocessing.AfterPlacementHook>` to subdivide cell
-populations into labelled subpopulations of same cell counts randomly.
+populations into labelled subpopulations randomly.
 The number of labels defines the number of subpopulations.
 
 * ``cell_types``: List of CellType of the population to subdivide, used as references;
 
-* ``same_size``: *(bool, optional)* If `True`, the population is divided into equally sized subsets.
-  Default: `False`.
-
 * ``labels``: List of labels to assign to each population respectively, defines also the number of subpopulations;
 
+* ``same_size``: *(bool, optional)* If `True`, the population is divided into equally sized subsets.
+  Default: `False`.
 
 :class:`LabelMicrozones<.placement.microzones.LabelMicrozones>`
 ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

--- a/tests/test_io_molecular.py
+++ b/tests/test_io_molecular.py
@@ -144,3 +144,73 @@ class TestIoMolecular(
             for from_, to_ in cs.load_connections().as_globals():
                 connected[from_[0], to_[0]] += 1
             self.assertAll(connected == len(all_pc))
+
+    def test_regression_68(self):
+        self.cfg.connectivity = {}
+        self.cfg.after_placement = dict(
+            label_cell=dict(
+                strategy="cerebellar_models.placement.microzones.LabelCells",
+                cell_types=["io", "pc"],
+                labels=["p", "m"],
+            ),
+        )
+
+        self.cfg.connectivity.add(
+            "mli_to_pc",
+            dict(
+                strategy="bsb.connectivity.AllToAll",
+                presynaptic=dict(cell_types=["mli", "mli2"]),
+                postsynaptic=dict(cell_types=["pc"]),
+            ),
+        )
+        self.cfg.connectivity.add(
+            "io_to_pc_p",
+            dict(
+                strategy="bsb.connectivity.AllToAll",
+                presynaptic=dict(cell_types=["io"], labels=["p"]),
+                postsynaptic=dict(cell_types=["pc"], labels=["p"]),
+            ),
+        )
+        self.cfg.connectivity.add(
+            "io_to_mli_p",
+            dict(
+                strategy="cerebellar_models.connectome.io_molecular.ConnectomeIO_MLI",
+                presynaptic=dict(cell_types=["io"], labels=["p"]),
+                postsynaptic=dict(cell_types=["mli", "mli2"]),
+                io_pc_connectivity=["io_to_pc_p"],
+                mli_pc_connectivity=["mli_to_pc"],
+                pre_cell_pc="pc",
+            ),
+        )
+        self.cfg.connectivity.add(
+            "io_to_pc_m",
+            dict(
+                strategy="bsb.connectivity.AllToAll",
+                presynaptic=dict(cell_types=["io"], labels=["m"]),
+                postsynaptic=dict(cell_types=["pc"], labels=["m"]),
+            ),
+        )
+        self.cfg.connectivity.add(
+            "io_to_mli_m",
+            dict(
+                strategy="cerebellar_models.connectome.io_molecular.ConnectomeIO_MLI",
+                presynaptic=dict(cell_types=["io"], labels=["m"]),
+                postsynaptic=dict(cell_types=["mli", "mli2"]),
+                io_pc_connectivity=["io_to_pc_m"],
+                mli_pc_connectivity=["mli_to_pc"],
+                pre_cell_pc="pc",
+            ),
+        )
+        self.network.compile(append=True, skip_placement=True)
+        ps_io = self.network.get_placement_set("io")
+        io_plus = ps_io.get_labelled(["p"])
+        io_minus = ps_io.get_labelled(["m"])
+        io_not_plus = set(ps_io.load_ids()) - set(io_plus)
+        self.assertEqual(io_not_plus, set(io_minus), "minus and plus set should not overlap")
+        io_mli_m = self.network.get_connectivity_set("io_to_mli_m_io_to_mli")
+        io_mli_p = self.network.get_connectivity_set("io_to_mli_p_io_to_mli")
+
+        io_m = io_mli_m.load_connections().all()[0][:, 0]
+        io_p = io_mli_p.load_connections().all()[0][:, 0]
+        self.assertAll(np.unique(io_m) == io_minus, "All io minus cells should be connected")
+        self.assertAll(np.unique(io_p) == io_plus, "All io minus cells should be connected")

--- a/tests/test_io_molecular.py
+++ b/tests/test_io_molecular.py
@@ -146,7 +146,6 @@ class TestIoMolecular(
             self.assertAll(connected == len(all_pc))
 
     def test_regression_68(self):
-        self.cfg.connectivity = {}
         self.cfg.after_placement = dict(
             label_cell=dict(
                 strategy="cerebellar_models.placement.microzones.LabelCells",
@@ -156,17 +155,28 @@ class TestIoMolecular(
         )
 
         self.cfg.connectivity.add(
-            "mli_to_pc",
+            "mli_to_pc_m",
             dict(
-                strategy="bsb.connectivity.AllToAll",
+                strategy="bsb.connectivity.FixedIndegree",
+                indegree=1,
                 presynaptic=dict(cell_types=["mli", "mli2"]),
-                postsynaptic=dict(cell_types=["pc"]),
+                postsynaptic=dict(cell_types=["pc"], labels=["m"]),
+            ),
+        )
+        self.cfg.connectivity.add(
+            "mli_to_pc_p",
+            dict(
+                strategy="bsb.connectivity.FixedIndegree",
+                indegree=1,
+                presynaptic=dict(cell_types=["mli", "mli2"]),
+                postsynaptic=dict(cell_types=["pc"], labels=["p"]),
             ),
         )
         self.cfg.connectivity.add(
             "io_to_pc_p",
             dict(
-                strategy="bsb.connectivity.AllToAll",
+                strategy="bsb.connectivity.FixedIndegree",
+                indegree=1,
                 presynaptic=dict(cell_types=["io"], labels=["p"]),
                 postsynaptic=dict(cell_types=["pc"], labels=["p"]),
             ),
@@ -178,14 +188,15 @@ class TestIoMolecular(
                 presynaptic=dict(cell_types=["io"], labels=["p"]),
                 postsynaptic=dict(cell_types=["mli", "mli2"]),
                 io_pc_connectivity=["io_to_pc_p"],
-                mli_pc_connectivity=["mli_to_pc"],
+                mli_pc_connectivity=["mli_to_pc_p"],
                 pre_cell_pc="pc",
             ),
         )
         self.cfg.connectivity.add(
             "io_to_pc_m",
             dict(
-                strategy="bsb.connectivity.AllToAll",
+                strategy="bsb.connectivity.FixedIndegree",
+                indegree=1,
                 presynaptic=dict(cell_types=["io"], labels=["m"]),
                 postsynaptic=dict(cell_types=["pc"], labels=["m"]),
             ),
@@ -197,20 +208,28 @@ class TestIoMolecular(
                 presynaptic=dict(cell_types=["io"], labels=["m"]),
                 postsynaptic=dict(cell_types=["mli", "mli2"]),
                 io_pc_connectivity=["io_to_pc_m"],
-                mli_pc_connectivity=["mli_to_pc"],
+                mli_pc_connectivity=["mli_to_pc_m"],
                 pre_cell_pc="pc",
             ),
         )
-        self.network.compile(append=True, skip_placement=True)
+        self.network.compile(redo=True, skip_placement=True)
         ps_io = self.network.get_placement_set("io")
         io_plus = ps_io.get_labelled(["p"])
         io_minus = ps_io.get_labelled(["m"])
         io_not_plus = set(ps_io.load_ids()) - set(io_plus)
         self.assertEqual(io_not_plus, set(io_minus), "minus and plus set should not overlap")
-        io_mli_m = self.network.get_connectivity_set("io_to_mli_m_io_to_mli")
-        io_mli_p = self.network.get_connectivity_set("io_to_mli_p_io_to_mli")
+        # io_mli_m = self.network.get_connectivity_set("io_to_mli_m_io_to_mli")
+        # io_mli_p = self.network.get_connectivity_set("io_to_mli_p_io_to_mli")
+        io_mli_m = self.network.get_connectivity_set("io_to_pc_m")
+        io_mli_p = self.network.get_connectivity_set("io_to_pc_p")
 
         io_m = io_mli_m.load_connections().all()[0][:, 0]
         io_p = io_mli_p.load_connections().all()[0][:, 0]
+        pc_m = io_mli_m.load_connections().all()[1][:, 0]
+        pc_p = io_mli_p.load_connections().all()[1][:, 0]
+        print(f"check minus: {np.unique(io_m)} - {io_minus} ")
+        print(f"check plus: {np.unique(io_p)} - {io_plus} ")
+        print(f"check pc minus: {np.unique(pc_m)}")
+        print(f"check pc plus: {np.unique(pc_p)}  ")
         self.assertAll(np.unique(io_m) == io_minus, "All io minus cells should be connected")
         self.assertAll(np.unique(io_p) == io_plus, "All io minus cells should be connected")

--- a/tests/test_io_molecular.py
+++ b/tests/test_io_molecular.py
@@ -214,12 +214,11 @@ class TestIoMolecular(
         )
         self.network.compile(redo=True, skip_placement=True)
         ps_io = self.network.get_placement_set("io")
+        ps_pc = self.network.get_placement_set("pc")
         io_plus = ps_io.get_labelled(["p"])
         io_minus = ps_io.get_labelled(["m"])
         io_not_plus = set(ps_io.load_ids()) - set(io_plus)
         self.assertEqual(io_not_plus, set(io_minus), "minus and plus set should not overlap")
-        # io_mli_m = self.network.get_connectivity_set("io_to_mli_m_io_to_mli")
-        # io_mli_p = self.network.get_connectivity_set("io_to_mli_p_io_to_mli")
         io_mli_m = self.network.get_connectivity_set("io_to_pc_m")
         io_mli_p = self.network.get_connectivity_set("io_to_pc_p")
 
@@ -227,9 +226,15 @@ class TestIoMolecular(
         io_p = io_mli_p.load_connections().all()[0][:, 0]
         pc_m = io_mli_m.load_connections().all()[1][:, 0]
         pc_p = io_mli_p.load_connections().all()[1][:, 0]
-        print(f"check minus: {np.unique(io_m)} - {io_minus} ")
-        print(f"check plus: {np.unique(io_p)} - {io_plus} ")
-        print(f"check pc minus: {np.unique(pc_m)}")
-        print(f"check pc plus: {np.unique(pc_p)}  ")
-        self.assertAll(np.unique(io_m) == io_minus, "All io minus cells should be connected")
-        self.assertAll(np.unique(io_p) == io_plus, "All io minus cells should be connected")
+        self.assertAll(
+            np.isin(np.unique(io_m), io_minus), "All connect IO should be IO minus cells"
+        )
+        self.assertAll(np.isin(np.unique(io_p), io_plus), "All connect IO should be IO plus cells")
+        self.assertAll(
+            np.unique(pc_m) == ps_pc.get_labelled(["m"]),
+            "All PC minus should be connected to IO minus cells",
+        )
+        self.assertAll(
+            np.unique(pc_p) == ps_pc.get_labelled(["p"]),
+            "All PC plus should be connected to IO plus cells",
+        )

--- a/tests/test_microzones.py
+++ b/tests/test_microzones.py
@@ -82,10 +82,10 @@ class TestMicrozones(
 
     def test_random_labels_same_size(self):
         self.cfg.after_placement["label_cell"] = dict(
-                strategy="cerebellar_models.placement.microzones.LabelCells",
-                cell_types=["io"],
-                labels=self.labels,
-                same_size=True,
+            strategy="cerebellar_models.placement.microzones.LabelCells",
+            cell_types=["io"],
+            labels=self.labels,
+            same_size=True,
         )
         self.scaffold = Scaffold(self.cfg, self.storage)
         self.scaffold.compile()
@@ -101,9 +101,9 @@ class TestMicrozones(
     def test_random_labels(self):
         self.cfg.cell_types["io"] = dict(spatial=dict(radius=2, count=1000))
         self.cfg.after_placement["label_cell"] = dict(
-                strategy="cerebellar_models.placement.microzones.LabelCells",
-                cell_types=["io"],
-                labels=self.labels,
+            strategy="cerebellar_models.placement.microzones.LabelCells",
+            cell_types=["io"],
+            labels=self.labels,
         )
         self.scaffold = Scaffold(self.cfg, self.storage)
         self.scaffold.compile()
@@ -115,8 +115,7 @@ class TestMicrozones(
         for label in unique_labels:
             filt = ps.get_labelled([label])
             self.assertLess(
-                np.abs(filt.size - len(ps) * 1/len(unique_labels))
-                / np.sqrt(filt.size),
+                np.abs(filt.size - len(ps) * 1 / len(unique_labels)) / np.sqrt(filt.size),
                 3.27,
                 "This test should fail only once in every 1000 trials",
             )

--- a/tests/test_microzones.py
+++ b/tests/test_microzones.py
@@ -79,3 +79,44 @@ class TestMicrozones(
         self.scaffold = Scaffold(self.cfg, self.storage)
         self.scaffold.compile()
         self._check_labels(axis=2)
+
+    def test_random_labels_same_size(self):
+        self.cfg.after_placement["label_cell"] = dict(
+                strategy="cerebellar_models.placement.microzones.LabelCells",
+                cell_types=["io"],
+                labels=self.labels,
+                same_size=True,
+        )
+        self.scaffold = Scaffold(self.cfg, self.storage)
+        self.scaffold.compile()
+        ps = self.scaffold.get_placement_set("io")
+        unique_labels = [
+            list(label)[0] for label in self.scaffold.get_placement_set("io").get_unique_labels()
+        ]
+        self.assertAll(np.isin(self.labels, unique_labels))
+        for label in unique_labels:
+            filt = ps.get_labelled([label])
+            self.assertEqual(filt.size, 8)
+
+    def test_random_labels(self):
+        self.cfg.cell_types["io"] = dict(spatial=dict(radius=2, count=1000))
+        self.cfg.after_placement["label_cell"] = dict(
+                strategy="cerebellar_models.placement.microzones.LabelCells",
+                cell_types=["io"],
+                labels=self.labels,
+        )
+        self.scaffold = Scaffold(self.cfg, self.storage)
+        self.scaffold.compile()
+        ps = self.scaffold.get_placement_set("io")
+        unique_labels = [
+            list(label)[0] for label in self.scaffold.get_placement_set("io").get_unique_labels()
+        ]
+        self.assertAll(np.isin(self.labels, unique_labels))
+        for label in unique_labels:
+            filt = ps.get_labelled([label])
+            self.assertLess(
+                np.abs(filt.size - len(ps) * 1/len(unique_labels))
+                / np.sqrt(filt.size),
+                3.27,
+                "This test should fail only once in every 1000 trials",
+            )

